### DISCUSSION TITLE
`using` statememt Duck-Type support

### DISCUSSION THREAD

**Discussion by toupswork** [dotnet/csharplang#8517] [Language Ideas] [answered]
Proposal: Allow using Statement for Duck-Typed Structs Without IDisposable to Avoid Boxing

Summary:
This proposal suggests allowing the using statement to work on structs that contain a Dispose() method without requiring them to implement IDisposable. The goal is to avoid boxing, which currently occurs when structs implement interfaces, while still leveraging the automatic disposal pattern.

Motivation:
In performance-critical scenarios, boxing a struct to call IDisposable.Dispose() results in heap allocations that could be avoided. The using statement is valuable for managing resources, but requiring IDisposable on structs forces boxing. Allowing a duck-typing approach, where structs with a Dispose() method can be used with using, would

**Comment by HaloFour** [comment, marked as answer]
See: #93

Also see: https://github.com/dotnet/csharplang/blob/main/meetings/2018/LDM-2018-12-12.md#when-we-start-recognizing-pattern-dispose-do-we-call-it-in-foreach

**Reply by toupswork** [reply]
I really appreciate that. thank you

**Comment by HaloFour** [comment]
> but requiring IDisposable on structs forces boxing

This is not true unless you cast the struct to the `IDisposable` interface.  The C# compiler emits a constrained method call which informs the runtime to not box.

Also, the compiler does support pattern-based `using` for `ref struct` since they can't implement interfaces, but it was decided to not extend that support beyond that one use case.

**Reply by toupswork** [reply]
thank you
